### ISSUE TITLE
Add ts eslint rule: explicit-function-return-type

### ISSUE THREAD

**Issue by zepumph** [phetsims/chipper#1215]
From https://github.com/phetsims/chipper/issues/1114, there are currently 2963 issues with default options (which I'm unsure we will want). We certainly want this rule though, so I'll take a look. 

To turn it on (not sure if these are the options either)

<details>

```diff
Index: eslint/.eslintrc.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/eslint/.eslintrc.js b/eslint/.eslintrc.js
--- a/eslint/.eslintrc.js	(revision 091109036fa7ad87accd58e81223959e287f2ae1)
+++ b/eslint/.eslintrc.js	(date 1649108358576)
@@ -40,7 +40,16 @@
         // Put TypeScript specific rules here
         // '@typescript-eslint/no-inferrable-types': 'error',
         // '@typescript-eslint/no-unnecessary-type-constraint': 'error',
-        '@typescript-eslint/member-delimiter-style': 'error' // semi colons in type declarations.
+        '@typescript-eslint/member-delimiter-style': 'error', // semi colons in type declarations.
+
+        // functions get return types.
+        '@typescript-eslint/explicit-function-return-type': [ 'error', {
+          allowExpressions: true,
+          allowTypedFunctionExpressions: false,
+          allowHigherOrderFunctions: true,
+          allowDirectConstAssertionInArrowFunctions: true,
+          allowConciseArrowFunctionExpressionsStartingWithVoid: false
+        } ]
 
         // Typescript rules that require type information (may be slow)
         // These require parserOptions.project.

**Comment by zepumph** [comment]
I'm not sure we are going to be able to use this rule, because it is requiring return types on all functions, not just methods. There is no option to check just method declarations. That said, if we adapted our own (not recommending) we could just check on function declarations and not on function/arrow function expressions.

**Comment by zepumph** [comment]
It looks like there is a compiler option: `noImplicitReturns`. If this does what we want (only for methods, not all anonymous functions). The doc for this is `/* Report error when not all code paths in function return a value. */`. So I don't think it will help, but it is already a lint rule I believe, so I'm going to turn it on.

**Comment by zepumph** [comment]
@jessegreenberg offered to help out on this.

**Comment by jessegreenberg** [comment]
Surprisingly nothing different had to be done for custom TS rules. I added a custom lint rule that requires a return type for method definitions excluding constructors getters and setters. 

There are 1584 errors, I scanned them and they seem like legit hits. I disabled it for now.

**Comment by samreid** [comment]
How difficult is it to write an autofix step that uses the inferred type?  I'm not saying to commit that blindly, but it could convert 1584 occurrences and then spot checking them could be much faster than writing all of them.

**Comment by jessegreenberg** [comment]
That could be worthwhile, Ill take a look.

**Comment by jessegreenberg** [comment]
Yes, this is working well! This rule now requires type information (`parserOptions.project` in .eslintrc.js and is slower). But running with `--fix` will insert the inferred type automatically.

**Comment by zepumph** [comment]
Can we do this and then remove the dependency on parserOptions.project after we fix?

**Comment by zepumph** [comment]
With the great help from @jessegreenberg I was able to turn this rule on! Thanks for helping to squash false positives. There should be no spots where the fixer input `any`. Instead I looked there myself to see what was best. Type checking and linting is passing, and I commented out the fixer so that we don't nee type information for the rule generally. Thanks @jessegreenberg for the doing the brunt of the work. What a great lintrule!
